utilities/conjure.1
.TH conjure 1 "Date: 2002/02/15 01:00:00" "ImageMagick"
.SH NAME
conjure - process a Magick Scripting Language (MSL) script

\fBWARNING: Conjure and MSL are in very early development and the
syntax is subject to change!\fP

.SH SYNOPSIS

\fBconjure\fP \fB[\fP \fIoptions\fP \fB]\fP \fIscript.msl\fP
\fB[ [\fP \fIoptions\fP \fB]\fP \fIscript.msl\fP \fB]\fP

.SH DESCRIPTION

The Magick scripting language (MSL) will primarily
benefit those that want to accomplish custom image processing tasks but
do not wish to program, or those that do not have access to a Perl
interpreter or a compiler.  The interpreter is called conjure and here
is an example script:


    <?xml version="1.0" encoding="UTF-8"?>
    <image size="400x400" >
      <read filename="image.gif" />
      <get width="base-width" height="base-height" />
      <resize geometry="%[dimensions]" />
      <get width="width" height="height" />
      <print output=
        "Image sized from %[base-width]x%[base-height]
         to %[width]x%[height].\\n" />
      <write filename="image.png" />
    </image>

invoked with


    conjure -dimensions 400x400 incantation.msl

All operations will closely follow the key/value pairs defined in
PerlMagick, unless otherwise noted.


\fBConjure\fP is in the early stages of development as of
April 2002.


This early announcement is to allow ImageMagick users to contribute ideas 
early in the process so when the scripting language is released it will
be useful/stable from the get-go!  If you want to contribute suggestions
about the Magick scripting language (MSL), post them to
magick-developers@imagemagick.org.

.SH OPTIONS

Options are processed in command line order. Any option you specify on
the command line remains in effect until it is explicitly changed by specifying
the option again with a different effect, or if it is changed by a statement
in the scripting language.


You can define your own keyword/value pairs on the command line.
The script can then use this information when setting values by including
%[keyword] in the string.  For example, if you included
"-dimensions 400x400" on the command line, as illustrated above,
then any string
containing "%[dimensions]" would have 400x400 substituted.
The "%[string]" can be used either an entire string, such as
geometry="%[dimensions]" or as a part of a string such as
filename="%[basename].png".

The keyword can be any string except for the following reserved
strings (in any upper, lower, or mixed case variant): \fBdebug\fP,
\fBhelp\fP, and \fBverbose\fP, whose usage is described below.

The value can be any string.  If
either the keyword or the value contains white space or any
symbols that have special meanings to your shell such as "#",
"|",
or
"%", enclose the string in quotation marks or use "\\" to escape the white
space and special symbols.

Keywords and values are case dependent.  "Key",
"key",
and "KEY" would
be three different keywords.


For a more detailed description of each option, see
\fIImageMagick(1)\fP.

.TP
.B "-debug \fI<event-mask>"\fP
\fRenable debug printout
.TP
.B "-help"
\fRprint usage instructions
.TP
.B "-verbose"
\fRprint detailed information about the image
.TP
.B "-version"
\fRprint ImageMagick version string
.SH MAGICK SCRIPTING LANGUAGE
The Magick Scripting Language (MSL) presently defines the following
elements and their attributes:
.in 15

.in 15
.B "<image>"
.in 20
 \fR
.in 20
Define a new image object.
\fB</image>\fP destroys it. Because of this, if you wish
to reference multiple "subimages" (aka pages or layers),
you can embed one \fBimage\fP element inside of
another. For example:

.in 15
.in 20


    <image>
    <read filename="input.png" />
    <get width="base-width" height="base-height" />
    <image height="base-height" width="base-width">
    <image />
    <write filename="output.mng" />
    </image>

.in 15
.in 20


    <image size="400x400" />

.in 15
.in 15
.B "<group>"
.in 20
 \fR
.in 20
Define a new group of image objects.
By default, images are only valid for the life of
their \fB<image>\fPelement.

.in 15
.in 20


    <image>   -- creates the image
    .....     -- do stuff with it
    </image>  -- dispose of the image

.in 15
.in 20

However, in a group, all images in that group will stay 
around for the life of the group:

.in 15
.in 20


    <group>                           -- start a group
        <image>                       -- create an image
        ....                          -- do stuff
        </image>                      -- NOOP
        <image>                       -- create another image
        ....                          -- do more stuff
        </image>                      -- NOOP
        <write filename="image.mng" />  -- output
    </group>                          -- dispose of both images

.in 15
.in 15
.B "<read>"
.in 20
 \fR
.in 20
Read a new image from a disk file.

.in 15
.in 20


    <read filename="image.gif" />

.in 15
.in 20
To read two images use

.in 15
.in 20


    <read filename="image.gif" />
    <read filename="image.png />

.in 15
.in 15
.B "<write>"
.in 20
 \fR
.in 20
Write the image(s) to disk, either as
a single multiple-image file or multiple ones if necessary.

.in 15
.in 20


     <write filename=image.tiff" />
.in 15
.B "<get>"
.in 20
 \fR
.in 20
Get any attribute recognized by
PerlMagick's GetAttribute() and stores it as an image attribute for later
use. Currently only \fIwidth\fP and \fIheight\fP are supported.
.in 15
.in 20


    <get width="base-width" height="base-height" />
    <print output="Image size is %[base-width]x%[base-height].\\n" />

.in 15
.in 15
.B "<set>"
.in 20
 \fR
.in 20
background, bordercolor, clip-mask, colorspace, density,
magick, mattecolor, opacity.  Set an attribute recognized by
PerlMagick's GetAttribute().
.in 15
.in 15
.B "<border>"
.in 20
 \fR
.in 20
    fill, geometry, height, width
.in 15
.in 15
.B "<blur>"
.in 20
 \fR
.in 20
    radius, sigma
.in 15
.in 15
.B "<charcoal>"
.in 20
 \fR
.in 20
    radius, sigma
.in 15
.in 15
.B "<chop>"
.in 20
 \fR
.in 20
    geometry, height, width, x, y
.in 15
.in 15
.B "<crop>"
.in 20
 \fR
.in 20
    geometry, height, width, x, y
.in 15
.in 15
.B "<despeckle>"
.in 20
 \fR
.in 15
.B "<emboss>"
.in 20
 \fR
.in 20
    radius, sigma
.in 15
.in 15
.B "<enhance>"
.in 20
 \fR
.in 15
.B "<equalize>"
.in 20
 \fR
.in 15
.B "<flip>"
.in 20
 \fR
.in 15
.B "<flop>"
.in 20
 \fR
.in 15
.B "<frame>"
.in 20
 \fR
.in 20
    fill, geometry, height, width, x, y, inner, outer
.in 15
.in 15
.B "<get>"
.in 20
 \fR
.in 20
    height, width
.in 15
.in 15
.B "<image>"
.in 20
 \fR
.in 20
    background, color, id, size
.in 15
.in 15
.B "<magnify>"
.in 20
 \fR
.in 15
.B "<minify>"
.in 20
 \fR
.in 15
.B "<normalize>"
.in 20
 \fR
.in 15
.B "<print>"
.in 20
 \fR
.in 20
    output
.in 15
.in 15
.B "<read>"
.in 20
 \fR
.in 15
.B "<resize>"
.in 20
 \fR
.in 20
    blur, filter, geometry, height, width
.in 15
.in 15
.B "<roll>"
.in 20
 \fR
.in 20
    geometry, x, y
.in 15
.in 15
.B "<rotate>"
.in 20
 \fR
.in 20
    degrees
.in 15
.in 15
.B "<sample>"
.in 20
 \fR
.in 20
    geometry, height, width
.in 15
.in 15
.B "<scale>"
.in 20
 \fR
.in 20
    geometry, height, width
.in 15
.in 15
.B "<sharpen>"
.in 20
 \fR
.in 20
    radius, sigma
.in 15
.in 15
.B "<shave>"
.in 20
 \fR
.in 20
    geometry, height, width
.in 15
.in 15
.B "<shear>"
.in 20
 \fR
.in 20
    x, y
.in 15
.in 15
.B "<solarize>"
.in 20
 \fR
.in 20
    threshold
.in 15
.in 15
.B "<spread>"
.in 20
 \fR
.in 20
    radius
.in 15
.in 15
.B "<stegano>"
.in 20
 \fR
.in 20
    image
.in 15
.in 15
.B "<stereo>"
.in 20
 \fR
.in 20
    image
.in 15
.in 15
.B "<swirl>"
.in 20
 \fR
.in 20
    degrees
.in 15
.in 15
.B "<texture>"
.in 20
 \fR
.in 20
    image
.in 15
.in 15
.B "<threshold>"
.in 20
 \fR
.in 20
    threshold
.in 15
.in 15
.B "<transparent>"
.in 20
 \fR
.in 20
    color
.in 15
.in 15
.B "<trim>"
.in 20
 \fR

.SH SEE ALSO

animate(1), display(1), composite(1), convert(1), identify(1),
import(1), mogrify(1), montage(1)

.SH COPYRIGHT

\fBCopyright (C) 2002 ImageMagick Studio\fP

\fBPermission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files ("ImageMagick"),
to deal in ImageMagick without restriction, including without limitation
the rights to use, copy, modify, merge, publish, distribute, sublicense,
and/or sell copies of ImageMagick, and to permit persons to whom the ImageMagick
is furnished to do so, subject to the following conditions:\fP

\fBThe above copyright notice and this permission notice shall be included
in all copies or substantial portions of ImageMagick.\fP

\fBThe software is provided "as is", without warranty of any kind, express
or implied, including but not limited to the warranties of merchantability,
fitness for a particular purpose and noninfringement.In no event shall
ImageMagick Studio be liable for any claim, damages or other liability,
whether in an action of contract, tort or otherwise, arising from, out
of or in connection with ImageMagick or the use or other dealings in
ImageMagick.\fP

\fBExcept as contained in this notice, the name of the
ImageMagick Studio LLC shall not be used in advertising or otherwise to
promote the sale, use or other dealings in ImageMagick without prior written
authorization from the ImageMagick Studio.\fP
.SH AUTHORS

\fI
John Cristy, ImageMagick Studio LLC,
.in 7
Glenn Randers-Pehrson, ImageMagick Studio LLC,
.in 7
Leonard Rosenthol, ImageMagick Studio LLC.
\fP



NEWS
This file was last updated on April 1, 2009.

Significant changes associated with GraphicsMagick 1.1.15 (April 1, 2009)

  Bugs Fixed:

    o The MagickLibVersion define overflowed as of 1.1.10 since only
      one digit was reserved for the minor version.  Fixed now.

    o MSL: Fix for handling of geometry x,y values in association with
      gravity.

    o TIFF: Ping mode is now fast like it should have been.

    o Conjure subcommand was returning inverted status to the
      environment.

    o Benchmark subcommand was leaking memory for each iteration.

    o Magick++ Image color quantization error measurement was
      performing a wasteful operation.

    o GraphicsMagick.spec was not installing module .la files by default.

    o Bounding box query was failing rather than returning a bounding box
      matching the entire image if there was no subimage to bound.

    o Memory leak/use fixes.

  Security Fixes:

    o Exclude all virtual delegates and coders for pseudo-formats when
      deciding to use file extension as the format.

    o Fix assertion thrown for BMP and DIB files with negative image
      height values.

Significant changes associated with GraphicsMagick 1.1.14 (released June 5, 2008)

  Bugs Fixed:

    o Noise generation was not working correctly for anything but a Q8 build.

    o Poisson noise generator was extremely slow.  Now it is reasonably fast.

    o PDF reader now computes bounding box for rotated PDFs correctly.

  Security Fixes:

    o Cineon reader: Fixed crash with broken file from Sami Liedes.

    o PICT reader: Fixed crash with broken files from Sami Liedes.

    o XCF reader: Fixed crash with broken files from Sami Liedes.

  Security Issue Not Fixed:

    o JP2/JPC reader: Several JPEG-2000 files from Sami Liedes broken collection
      cause crashes in Jasper.  This is not GraphicsMagick's fault.

Significant changes associated with GraphicsMagick 1.1.13 (released May 28, 2008)

  Bugs Fixed:

    o Documentation is now installed according to the conventions established by
      the configure script.  This results in documentation being installed under
      /usr/local/share/doc/GraphicsMagick by default.

    o HWB colorspace now works correctly.

    o Composition with CopyOpacity now produces an image with
      transparency (as expected).

    o Composition now preserves the canvas colorspace.

    o Composition with a displacement map (-displace) no longer leaks an image.

    o Alpha composition now works as expected when both pixels involved
      include transparency.

    o -gamma multiple channel syntax now works as documented.

    o Now compiles with Visual C++ 2008.

Significant changes associated with GraphicsMagick 1.1.12 (released April 28, 2008)

  Security Fixes:

    o Do not access X11 or invoke convenience or stealth delegate
      programs based on the file extension. In particular, these file
      extensions are rejected for consideration as a format specifier:
      'autotrace', 'browse', 'dcraw', 'edit', 'gs-color',
      'gs-color+alpha', 'gs-gray', 'gs-mono', 'launch', 'mpeg-encode',
      'print', 'scan', 'show', 'win', 'xc', and 'x'.

  Bugs Fixed:

    o magick/effect.c: Should now compile for ARM CPU.

    o TIFF: Don't request Kodak private tags since these cause some
      versions of libtiff to misbehave.

    o When performing string expansion of image attribute identifiers, skip
      those which require access to image pixels if image pixels are not
      present.

    o CropImageToHBITMAP(), ImageToHBITMAP(): Fix leak of bitmap handle.

Significant changes associated with GraphicsMagick 1.1.11 (released February 23, 2008)

  Bugs Fixed:

    o BMP: Support large files.

    o DIB: Support large files.

    o PNG: Fix depth handling with 16-bit PNG files in the Q8 build.

    o SUN: Properly report image depth.

    o TIFF: Endian option (-endian) now controls TIFF byte endian order
      rather than bit fill order.

    o DCM, DIB, XBM, XCF, XWD: Eliminate integer overflow vulnerability
      (IDefense 09.19.07).

    o HSL colorspace transform: Avoid optimization bug noticed on Opteron
      with GCC.

    o HWB colorspace transform: Avoid optimization bug noticed on Opteron
      with GCC.

    o RGBTransformImage()/TransformRGBImage(): Was using HWB colorspace
      when HSL was requested.

    o Successfully reads files with names like 'file[123]'.

    o 'gm display': No longer rely on isatty() to determine if input is
      from a pipe (use 'gm display -' to display an image read from a pipe).

  Feature Improvements:

    o 'identify +ping' forces the pixels to be read (similar to GM 1.2).

    o 'gm -version' now indicates if build supports "Large Memory" (i.e. 64-bit).

    o TIFF: Use '-define tiff:fill-order={msb2lsb|lsb2msb}' to control
      TIFF bit fill order.


  Performance Improvements:

    o No longer bogs down if a directory contains hundreds of thousands of
      files and the filename looks like a wildcard specification.


Significant changes associated with GraphicsMagick 1.1.10 (released September 19, 2007)

  Bugs Fixed:

    o Image rotate by -90/270 degrees was producing wrong output.

Significant changes associated with GraphicsMagick 1.1.9 (released September 16, 2007)

  Bugs Fixed:

    o In mogrify command, don't remove file name based on random junk in
      memory.
    o Fixed memory leak when reading MPC files.
    o Fixed crash when writing MIFF format and depth is not expected 8/16/32/.
    o Find latest Ghostscript under Windows which identifies itself as
      "GPL Ghostscript".
    o In mogrify command, don't leak memory in the case where the image
      file contains multiple frames.
    o Fixed crash in PNG and JPEG coders when the image to be written is
      part of an image list.
    o PNG reader errors are now properly reported to the user.
    o TIFF output can now be written to a pipe or other non-seekable
      destination.

utilities/conjure.c
/*
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%                                                                             %
%                                                                             %
%                                                                             %
%                CCCC   OOO   N   N  JJJJJ  U   U  RRRR   EEEEE               %
%               C      O   O  NN  N    J    U   U  R   R  E                   %
%               C      O   O  N N N    J    U   U  RRRR   EEE                 %
%               C      O   O  N  NN  J J    U   U  R R    E                   %
%                CCCC   OOO   N   N  JJJ     UUU   R  R   EEEEE               %
%                                                                             %
%                     Interpret Magick Scripting Language.                    %
%                                                                             %
%                                                                             %
%                              Software Design                                %
%                                John Cristy                                  %
%                               December 2001                                 %
%                                                                             %
%                                                                             %
%  Copyright (C) 2002 ImageMagick Studio, a non-profit organization dedicated %
%  to making software imaging solutions freely available.                     %
%                                                                             %
%  Permission is hereby granted, free of charge, to any person obtaining a    %
%  copy of this software and associated documentation files ("ImageMagick"),  %
%  to deal in ImageMagick without restriction, including without limitation   %
%  the rights to use, copy, modify, merge, publish, distribute, sublicense,   %
%  and/or sell copies of ImageMagick, and to permit persons to whom the       %
%  ImageMagick is furnished to do so, subject to the following conditions:    %
%                                                                             %
%  The above copyright notice and this permission notice shall be included in %
%  all copies or substantial portions of ImageMagick.                         %
%                                                                             %
%  The software is provided "as is", without warranty of any kind, express or %
%  implied, including but not limited to the warranties of merchantability,   %
%  fitness for a particular purpose and noninfringement.  In no event shall   %
%  ImageMagick Studio be liable for any claim, damages or other liability,    %
%  whether in an action of contract, tort or otherwise, arising from, out of  %
%  or in connection with ImageMagick or the use or other dealings in          %
%  ImageMagick.                                                               %
%                                                                             %
%  Except as contained in this notice, the name of the ImageMagick Studio     %
%  shall not be used in advertising or otherwise to promote the sale, use or  %
%  other dealings in ImageMagick without prior written authorization from the %
%  ImageMagick Studio.                                                        %
%                                                                             %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%  Conjure interprets and executes scripts in the Magick Scripting Language
%  (MSL). The Magick scripting language (MSL) will primarily benefit those
%  that want to accomplish custom image processing tasks but do not wish
%  to program, or those that do not have access to a Perl interpreter or a
%  compiler. The interpreter is called conjure and here is an example script:
%
%      <?xml version="1.0" encoding="UTF-8"?>
%      <image size="400x400" >
%      <read filename="image.gif" />
%      <get width="base-width" height="base-height" />
%      <resize geometry="%[dimensions]" />
%      <get width="width" height="height" />
%      <print output="Image sized from %[base-width]x%[base-height]
%          to %[width]x%[height].\n" />
%      <write filename="image.png" />
%      </image>
%
%
*/

/*
  Include declarations.
*/
#include "studio.h"
#include "attribute.h"
#include "command.h"
#include "log.h"
#include "version.h"
#include "utility.h"

/*
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%                                                                             %
%                                                                             %
%                                                                             %
%   C o n j u r e U s a g e                                                   %
%                                                                             %
%                                                                             %
%                                                                             %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%  ConjureUsage() displays the program command syntax.
%
%  The format of the ConjureUsage method is:
%
%      void ConjureUsage()
%
*/
static void ConjureUsage(void)
{
  static const char
    *options[]=
    {
      "-debug events        display copious debugging information",
      "-help                print program options",
      "-verbose             print detailed information about the image",
      "-version             print version information",
      (char *) NULL
    };

  const char
    **p;

  (void) printf("Version: %.1024s\n",GetMagickVersion((unsigned long *) NULL));
  (void) printf("Copyright: %.1024s\n\n",GetMagickCopyright());
  (void) printf("Usage: %.1024s [options ...] file [ [options ...] file ...]\n",
    SetClientName((char *) NULL));
  (void) printf("\nWhere options include:\n");
  for (p=options; *p != (char *) NULL; p++)
    (void) printf("  %.1024s\n",*p);
  (void) printf("\nIn additiion, define any key value pairs required by "
    "your script.  For\nexample,\n\n");
  (void) printf("    conjure -size 100x100 -color blue -foo bar script.msl\n");
  Exit(0);
}

/*
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%                                                                             %
%                                                                             %
%                                                                             %
%  M a i n                                                                    %
%                                                                             %
%                                                                             %
%                                                                             %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%
*/
int main(int argc,char **argv)
{
  char
    *option;

  ExceptionInfo
    exception;

  Image
    *image;

  ImageInfo
    *image_info;

  register long
    i;

  unsigned int
    status;

  InitializeMagick(*argv);
  ReadCommandlLine(argc,&argv);
  status=ExpandFilenames(&argc,&argv);
  if (status == False)
    MagickFatalError(ResourceLimitFatalError,"MemoryAllocationFailed",
      (char *) NULL);
  if (argc < 2)
    ConjureUsage();
  GetExceptionInfo(&exception);
  image_info=CloneImageInfo((ImageInfo *) NULL);
  image_info->attributes=AllocateImage(image_info);
  status=True;
  for (i=1; i < argc; i++)
  {
    option=argv[i];
    if ((strlen(option) != 1) && ((*option == '-') || (*option == '+')))
      {
        if (LocaleCompare("debug",option+1) == 0)
          {
            if (*option == '-')
              {
                i++;
                if (i == argc)
                  MagickFatalError(OptionFatalError,"MissingEventMask",
                    option);
                (void) SetLogEventMask(argv[i]);
              }
            continue;
          }
        if (LocaleCompare("help",option+1) == 0)
          {
            if (*option == '-')
              ConjureUsage();
            continue;
          }
        if (LocaleCompare("verbose",option+1) == 0)
          {
            image_info->verbose=(*option == '-');
            continue;
          }
        if (LocaleCompare("version",option+1) == 0)
          {
            (void) fprintf(stdout,"Version: %.1024s\n",
              GetMagickVersion((unsigned long *) NULL));
            (void) fprintf(stdout,"Copyright: %.1024s\n\n",
              GetMagickCopyright());
            Exit(0);
            continue;
          }
        /*
          Persist key/value pair.
        */
        (void) SetImageAttribute(image_info->attributes,option+1,(char *) NULL);
        status&=SetImageAttribute(image_info->attributes,option+1,argv[i+1]);
        if (status == False)
          MagickFatalError(ImageFatalError,"UnableToPersistKey",option);
        i++;
        continue;
      }
    /*
      Interpret MSL script.
    */
    (void) SetImageAttribute(image_info->attributes,"filename",(char *) NULL);
    status&=SetImageAttribute(image_info->attributes,"filename",argv[i]);
    if (status == False)
      MagickFatalError(ImageFatalError,"UnableToPersistKey",argv[i]);
    (void) FormatString(image_info->filename,"msl:%.1024s",argv[i]);
    image=ReadImage(image_info,&exception);
    if (exception.severity != UndefinedException)
      CatchException(&exception);
    status&=image != (Image *) NULL;
    if (image != (Image *) NULL)
      DestroyImageList(image);
  }
  DestroyImageInfo(image_info);
  DestroyMagick();
  LiberateMemory((void **) &argv);
  Exit(!status);
  return(False);
}
